### ISSUE TITLE
OutOfSync when deployment has topologySpreadConstraints and resource limits 

### ISSUE THREAD

**Issue by imeliran** [argoproj/argo-cd#16400]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**
We noticed that deployments that are a part of the Application become OutOfSync if they have topologySpreadConstraints as part of specs . when looking at the diff pane it shows unrelated diff regarding the resource limits:
![image](https://github.com/argoproj/argo-cd/assets/36511676/566a258b-e593-45bc-9284-9dca93f9fe9d)

**To Reproduce**
Create argo application that deploy the below manifest:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: devops
  name: petclinic-demo-app-1
  labels:
    app: petclinic-demo-app-1
spec:
  replicas: 1
  selector:
    matchLabels:
      app: petclinic-demo-app-1
  template:
    metadata:
      labels:
        app: petclinic-demo-app-1
    spec:
      topologySpreadConstraints:
        - maxSkew: 1
          minDomains: 1
          topologyKey: "kubernetes.io/hostname"
          whenUnsatisfiable: "DoNotSchedule"
          labelSelector:
            matchLabels:
              app: petclinic-demo-app-1
          matchLabelKeys:
            - pod-template-hash
          nodeAffinityPolicy: Honor
          nodeTaintsPolicy: Ignore
      containers:
        - name: app
          image: jbrisbin/spring-petclinic
          env:
          - name: "_JAVA_OPTIONS"
            value: "-Xmx1G -Xms1G -XX:MaxDirectMemorySize=64m"
          resources:
            limits:
              cpu: 1000m
              memory: 1Gi
            requests:
              cpu: 1000m
              memory: 1Gi
```
And you should see app outofsync with a diff of limits.
2) Remove the topologySpreadConstraints from deployment :
```yaml
      topologySpreadConstraints:
        - maxSkew: 1
          minDomains: 1
          topologyKey: "kubernetes.io/hostname"
          whenUnsatisfiable: "DoNotSchedule"
          labelSelector:
            matchLabels:
              app: petclinic-demo-app-1
          matchLabelKeys:
            - pod-template-hash
          nodeAffinityPolicy: Honor
          nodeTaintsPolicy: Ignore
```
And you should see that outofsync is no longer seen.

**Expected behavior**
App should be synced properly in terms of resource limits regardless the existence of topologySpreadConstraints

**Screenshots**

**Version**

```shell
v2.9.1+58b04e5
```

**Logs**

```
Paste any relevant application logs here.
```

**Comment by ashutosh16** [comment]
I have seen a similar issue in one environment where resource limits diffing, however, in another environment with an identical configuration there is no difference. The spec doesn't contain any extra fields like  `topologySpreadConstraints`
```
resources:
            limits:
              cpu: '4'
              memory: 2Gi
            requests:
              cpu: 500m
              memory: 2Gi
```
![Screenshot 2024-01-08 at 2 34 45 PM](https://github.com/argoproj/argo-cd/assets/11219262/c94e80b6-c679-4194-b773-20b475b989b0)

**Comment by rwong2888** [comment]
I am getting the same issue.

If I remove `matchLabelKeys` from `topologySpreadConstraints`, it reverts back to the regular behaviour.

```
matchLabelKeys:
- pod-template-hash
```

**Comment by tooptoop4** [comment]
@imeliran @ashutosh16 @rwong2888 what k8s version are u using? i wonder if this relates to https://github.com/argoproj/argo-cd/issues/15176

**Comment by rwong2888** [comment]
I am on 1.27 @tooptoop4. I suspect the same. Was discussing with @crenshaw-dev and maybe in 2.11 it will get resolved.

**Comment by tianqi-z** [comment]
I was on Kubernetes 1.27,  both matchLabelKeys and nodeTaintsPolicy is causing outofsync.
 The diff is about empty or default fields like:
 `nodeSelector: {}`
 `tolerations: []`
 `hostNetwork: false`
 .etc.
 Argo version 2.7.7

**Comment by rob-whittle** [comment]
> I have seen a similar issue in one environment where resource limits diffing, however, in another environment with an identical configuration there is no difference. The spec doesn't contain any extra fields like `topologySpreadConstraints`
> 
> ```
> resources:
>             limits:
>               cpu: '4'
>               memory: 2Gi
>             requests:
>               cpu: 500m
>               memory: 2Gi
> ```
> 
> ![Screenshot 2024-01-08 at 2 34 45 PM](https://private-user-images.githubusercontent.com/11219262/295052634-c94e80b6-c679-4194-b773-20b475b989b0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MDg5MjAsIm5iZiI6MTcwODUwODYyMCwicGF0aCI6Ii8xMTIxOTI2Mi8yOTUwNTI2MzQtYzk0ZTgwYjYtYzY3OS00MTk0LWI3NzMtMjBiNDc1Yjk4OWIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDA5NDM0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2YWYwMGJmYjFhODYwMmIxYjAyOGUxMjBmNjhlMzllOWU5YWRkNTlhY2ZhYTRhMjExODQyN2U5NDJkOWNjZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bpKlf_9J5PxXEv6P_WTFSVz0O4uCudvE5K_nd4BCqKc)

We are also seeing this issue on one of our kubernetes clusters.  Working fine on over environments with identical configuration though.

<img width="1228" alt="image" src="https://github.com/argoproj/argo-cd/assets/69803842/ce156f0e-cb25-458d-b86c-f0d2f23315b1">

**Comment by adam-harries-hub** [comment]
My team is also hitting this bug when including `topologySpreadConstraints` in the spec and using whole CPU values for requests/limits.

**Comment by henryzhao95** [comment]
Also having this issue on v2.9.5+f943664, where the sync was happy when the desired manifest has `resources.limits.cpu: 1000m` and the live manifest has `resources.limits.cpu: '1'`. But something made it unhappy - not sure if it was the addition of the `nodeTaintsPolicy` to our topology spread constraints.

**Comment by emmahsax** [comment]
Same issue on 2.7.9. The issue is with any field that's still in beta, so that includes `nodeAffinityPolicy`, `nodeTaintsPolicy`, and `matchLabelKeys`.

Hopefully they resolve this soon!

**Comment by pankajkumar911** [comment]
Looks like the same issue persists on Kubernetes version 1.29 
The flag matchLabelKeys is still in beta according to feature gates page.

![image](https://github.com/argoproj/argo-cd/assets/25081199/457c1ce1-697e-4269-93d2-a080da2ae7e4)

**Comment by easterbrennan** [comment]
Hitting a similar issue here with k8s 1.28 and argo v2.7.14.

Initially hit a problem with setting `InitialDelaySeconds: 0` on probes and now hitting similar resource diffs

![image](https://github.com/user-attachments/assets/456fdb36-1e9c-4b69-86dd-d845685baef0)

**Comment by RLProteus** [comment]
Can confirm I'm seeing this as well when adding `topologySpreadConstraints` to a deployment. Seeing both the `InitialDelaySeconds: 0` issue as well as resource request/limits marshaling failures on values that worked previously. 

ArgoCD v2.9.6 and K8s 1.28

**Comment by SavaMihai** [comment]
Hello, I'm having the same issue when using `topologySpreadConstraints` all the apps go Out Of Sync ...

**Comment by benniwiit** [comment]
Hi, we are facing the same issue in v2.11.4+e1284e1. We need to calculate the CPU limit based on the request, so we need integer. ArgoCD then always shows OutOfSync:
![image](https://github.com/user-attachments/assets/ca3d685c-c0cf-4595-9708-14eec390c376)

**Comment by jinleileiking** [comment]
workaround
<img width="440" alt="image" src="https://github.com/user-attachments/assets/84c5c38b-59e3-4353-adb9-d160360fa706">

**Comment by benniwiit** [comment]
We are facing this in both ways =).

<img width="750" alt="Screenshot 2024-10-17 at 06 36 03" src="https://github.com/user-attachments/assets/4ffb1fae-d2ff-45aa-99d2-9fb84e4c99aa">

<img width="989" alt="Screenshot 2024-10-17 at 06 39 04" src="https://github.com/user-attachments/assets/89ba8dbd-165b-4a3c-923b-61e94268c649">

**Comment by andrii-korotkov-verkada** [comment]
You can configure known types following https://argo-cd.readthedocs.io/en/stable/user-guide/diffing/#known-kubernetes-types-in-crds-resource-limits-volume-mounts-etc to deal with this. I'll close this, but feel free to re-open if this is not enough.

**Comment by rwong2888** [comment]
Following up on this, it appears my issue is resolved, at least in 2.13.0. Maybe, even earlier.

**Comment by purkhusid** [comment]
I'm encountering this in 2.14.2 with server side diff enabled. I managed work around this by adding the following to the ArgoCD config:
```
      resource.customizations.knownTypeFields.apps_Deployment: |
        - field: spec.template.spec.affinity
          type: core/v1/Affinity
```

Not sure if this issue should be closed since there seems to be a configuration that still has this failure.
